### PR TITLE
[AMDGPU] Implement 'llvm.get.fpenv' and 'llvm.set.fpenv'

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
This patch implements the LLVM floating point environment control
intrinsics and also exposes it through clang. We encode the floating
point environment as a 64-bit value that simply concatenates the values
of the mode registers and the current trap status. We only fetch the
bits relevant for floating point instructions. That is, rounding mode,
denormalization mode, ieee, dx10 clamp, debug, enabled traps, f16
overflow, and active exceptions.

### CODE REVIEW

**Comment by jhuber6** [comment]
Note that this patch is not quite ready to land. I encountered issues when working with `s_setreg`. The listing of `SOPK` instructions should have this as an instruction that takes a 16-bit zero extended immediate value. However, this was apparently not the case for the `s_setreg` node despite the fact that it was set for the `s_getreg` node. I made the necessary changes to consider this zero extended which makes it work in my case. However, this has exposed some failures in tests mostly relating to flat storage. These tests now fail due to being given a negative immediate value. I'm unsure where the necessary changes need to go to resolve that.

File: llvm/lib/Target/AMDGPU/SOPInstructions.td [hunk 1]
```diff
@@ -1122,7 +1122,7 @@ class S_SETREG_B32_Pseudo <list<dag> pattern=[]> : SOPK_Pseudo <
   pattern>;
 
 def S_SETREG_B32 : S_SETREG_B32_Pseudo <
-  [(int_amdgcn_s_setreg (i32 SIMM16bit:$simm16), i32:$sdst)]> {
+  [(int_amdgcn_s_setreg (i32 timm:$simm16), i32:$sdst)]> {
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SOPInstructions.td:1125]: This just reverts my patch https://github.com/llvm/llvm-project/pull/77997 and reintroduces the original problem.

Comment (jhuber6) [reply llvm/lib/Target/AMDGPU/SOPInstructions.td:1125]: Okay, so it's expected to be negative even though `s_getreg` isn't? We can keep it, but the problem I was having was that in global-isel it would error out constantly with the current usage.

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SOPInstructions.td:1125]: It is not expected to be negative, the original problem was that we used to force users to use negative constants. Now we can accept something like 0xf000 instead of a negative value.

Comment (jhuber6) [reply llvm/lib/Target/AMDGPU/SOPInstructions.td:1125]: I hit the `SOPK` error when trying to emit this in `globalIsel` for the `TRAPSTS` register. 
```
> llc -mtriple=amdgcn -mcpu=tahiti ../llvm/test/CodeGen/AMDGPU/llvm.fpenv.ll -global-isel
error: Illegal instruction detected: invalid immediate for SOPK instruction
S_SETREG_B32 killed renamable $sgpr4, 45057, implicit-def $mode, implicit $mode
```
I tried casting them so they would get sign extended, but I got similar issues.

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SOPInstructions.td:1125]: If it is sign extended, it should work.

Comment (jhuber6) [reply llvm/lib/Target/AMDGPU/SOPInstructions.td]: Okay, I think I got it to work by casting it to `int16_t` only on the store instruction. My previous mistake seems to have been changing the constant for both `set` and `get. It passes all the AMDGPU codegen tests now.

File: clang/include/clang/Basic/BuiltinsAMDGPU.def [hunk 2]
```diff
@@ -325,6 +325,9 @@ BUILTIN(__builtin_amdgcn_read_exec_hi, "Ui", "nc")
 
 BUILTIN(__builtin_amdgcn_endpgm, "v", "nr")
 
+BUILTIN(__builtin_amdgcn_get_fpenv, "WUi", "n")
```
Comment (arsenm) [inline clang/include/clang/Basic/BuiltinsAMDGPU.def:328]: @spavloff was there supposed to be a generic builtin for these?

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 3]
```diff
@@ -905,6 +905,9 @@ AMDGPULegalizerInfo::AMDGPULegalizerInfo(const GCNSubtarget &ST_,
   getActionDefinitionsBuilder(G_STACKRESTORE)
     .legalFor({PrivatePtr});
 
+  getActionDefinitionsBuilder(G_GET_FPENV).customFor({S64});
+  getActionDefinitionsBuilder(G_SET_FPENV).customFor({S64});
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:909]:
```suggestion
  getActionDefinitionsBuilder({G_GET_FPENV, G_SET_FPENV}).customFor({S64});
```

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 4]
```diff
@@ -4079,6 +4081,70 @@ SDValue SITargetLowering::lowerFP_EXTEND(SDValue Op, SelectionDAG &DAG) const {
...
+
+  SDVTList VTList = DAG.getVTList(MVT::i32, MVT::Other);
+  SDValue IntrinID =
+      DAG.getTargetConstant(Intrinsic::amdgcn_s_getreg, SL, MVT::i32);
+  SDValue GetModeReg = DAG.getNode(ISD::INTRINSIC_W_CHAIN, SL, VTList,
+                                   Op.getOperand(0), IntrinID, ModeHwRegImm);
+  SDValue GetTrapReg = DAG.getNode(ISD::INTRINSIC_W_CHAIN, SL, VTList,
+                                   Op.getOperand(0), IntrinID, TrapHwRegImm);
+  SDValue MergedReg =
+      DAG.getMergeValues({GetModeReg.getValue(1), GetTrapReg.getValue(1)}, SL);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:4103]: This should create a TokenFactor, not a merge

File: llvm/test/CodeGen/AMDGPU/llvm.fpenv.ll [hunk 5]
```diff
@@ -0,0 +1,114 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=amdgcn -mcpu=tahiti < %s | FileCheck -check-prefixes=GFX6 %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/llvm.fpenv.ll:2]: use an explicit -global-isel=0 for the dag cases

File: llvm/test/CodeGen/AMDGPU/llvm.fpenv.ll [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+; GFX11-NEXT:    v_dual_mov_b32 v0, s0 :: v_dual_mov_b32 v1, s1
+; GFX11-NEXT:    s_setpc_b64 s[30:31]
+entry:
+  %0 = tail call i64 @llvm.get.fpenv.i64()
+  ret i64 %0
+}
+
+declare void @llvm.set.fpenv.i64(i64)
+
+define void @set_fpenv(i64 %env) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/llvm.fpenv.ll:66]: should test with SGPR, and constant sources. also setenv from getenv

File: clang/test/CodeGenOpenCL/builtins-amdgcn.cl [hunk 7]
```diff
@@ -839,6 +839,18 @@ unsigned test_wavefrontsize() {
   return __builtin_amdgcn_wavefrontsize();
 }
 
+// CHECK-LABEL test_get_fpenv(
```
Comment (arsenm) [inline clang/test/CodeGenOpenCL/builtins-amdgcn.cl:842]: Ideally we would also warn if you used these without fenv access enabled

Comment (jhuber6) [reply clang/test/CodeGenOpenCL/builtins-amdgcn.cl:842]: Is this related to the DX10 clamp / traps potentially being disabled? Or is this an LLVM concept.

Comment (arsenm) [reply clang/test/CodeGenOpenCL/builtins-amdgcn.cl:842]: It's a standard C concept. You need to enable #pragma STDC FENV_ACCESS ON to do anything valid with the floating point environment

Comment (jhuber6) [reply clang/test/CodeGenOpenCL/builtins-amdgcn.cl:842]: Hm, I'm not sure. I feel like this is just letting the user access the hardware directly which has a different use-case to the end user.

Comment (arsenm) [reply clang/test/CodeGenOpenCL/builtins-amdgcn.cl:842]: This interface is 1 level removed from the hardware, but this cannot be legitimately used without fenv access enabled exactly the same way. It is not a different use case at all

Comment (jhuber6) [reply clang/test/CodeGenOpenCL/builtins-amdgcn.cl:842]: Added a sema check.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 8]
```diff
@@ -6826,6 +6826,10 @@ def warn_floatingpoint_eq : Warning<
   "comparing floating point with == or != is unsafe">,
   InGroup<DiagGroup<"float-equal">>, DefaultIgnore;
 
+def warn_fenv_access : Warning<
+  "floating point environment access without #pragma STDC FENV_ACCESS set ON">,
+  InGroup<DiagGroup<"float-env">>;
```
Comment (arsenm) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:6831]: this should be a separate patch (and ideally would cover the range of standard fenv access functions)

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 9]
```diff
@@ -4079,6 +4081,71 @@ SDValue SITargetLowering::lowerFP_EXTEND(SDValue Op, SelectionDAG &DAG) const {
...
+  SDValue CvtPtr =
+      DAG.getNode(ISD::BUILD_VECTOR, SL, MVT::v2i32, GetModeReg, GetTrapReg);
+  SDValue Result = DAG.getNode(ISD::BITCAST, SL, MVT::i64, CvtPtr);
+
+  return DAG.getMergeValues({Result, TokenReg}, SL);
+}
+
+SDValue SITargetLowering::lowerSET_FPENV(SDValue Op, SelectionDAG &DAG) const {
+  SDLoc SL(Op);
+  assert(Op.getOperand(1).getValueType() == MVT::i64);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:4115]: shouldn't assert on this. should either do zext-or-trunc, or return to fail legalize

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 10]
```diff
@@ -6940,6 +6946,42 @@ bool AMDGPULegalizerInfo::legalizeWaveID(MachineInstr &MI,
...
 
+static constexpr unsigned FPEnvModeBitField =
+    AMDGPU::Hwreg::HwregEncoding::encode(AMDGPU::Hwreg::ID_MODE, 0, 23);
+
+static constexpr unsigned FPEnvTrapBitField =
+    AMDGPU::Hwreg::HwregEncoding::encode(AMDGPU::Hwreg::ID_TRAPSTS, 0, 5);
+
+bool AMDGPULegalizerInfo::legalizeGetFPEnv(MachineInstr &MI,
+                                           MachineIRBuilder &B) const {
+  auto ModeReg =
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:6957]: should guard against unexpected types like the dAG case

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 11]
```diff
@@ -6940,6 +6946,42 @@ bool AMDGPULegalizerInfo::legalizeWaveID(MachineInstr &MI,
...
+                       /*HasSideEffects=*/true, /*isConvergent=*/false)
+          .addImm(FPEnvTrapBitField);
+  B.buildMergeLikeInstr(MI.getOperand(0).getReg(), {ModeReg, TrapReg});
+  MI.eraseFromParent();
+  return true;
+}
+
+bool AMDGPULegalizerInfo::legalizeSetFPEnv(MachineInstr &MI,
+                                           MachineIRBuilder &B) const {
+  auto Unmerge = B.buildUnmerge({S32, S32}, MI.getOperand(0));
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:6972]: Ditto, should guard against unexpected types

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 8 above)
Comment (jhuber6) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:6831]: There's no immediate users until this patch as far as I'm aware. Unless you're talking about expanding it to cover some other cases I'm not aware of. I guess I could split out the clang / llvm stuff but this was generally more convenient.

Comment (arsenm) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:6831]: This should cover every function in fenv access. This low level implementation detail of those functions is less interesting to report on than the actual user facing functions

Comment (jhuber6) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:6831]: I'm not specifically knowledgeable about how this should be handled in the grander scheme. Could I get rid of the warning and leave that for a separate patch?

Comment (arsenm) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: It must be a separate patch

**Review by arsenm — APPROVED** [review]
Don't really love adding the new not-generic builtin for this but OK I guess
